### ISSUE TITLE
[libc++] Array return value should be unique

### ISSUE THREAD

**Issue by rnicholl-google** [llvm/llvm-project#75654]
The return values of `std::array<T,0>::begin` and `std::array<T,0>::end` should be unique.

I drafted a PR to fix this, but it was blocked by a bug in the deletion of union destructors that needs to be fixed.

I will note: this is actually possibly to implement the standard behavior, but requires resolution of https://github.com/llvm/llvm-project/issues/56814 first

**Comment by frederick-vs-ja** [comment]
The "non-unique" return values are intended, see https://reviews.llvm.org/D80821.

As said by @ldionne:
> The Standard is currently unimplementable. We have to pick between:
> 
> 1. Not implementing constexpr support properly in `std::array<T, 0>`
> 2. Making `std::array<T, 0>` non-trivial even when `T` is trivial
> 3. Returning nullptr from `std::array<T, 0>::begin()`
> Libc++ initially picked (1). In [77b9abfc8e89](https://reviews.llvm.org/rG77b9abfc8e89ca627e4f9a1cc206bea131db6db1), we started implementing constexpr properly, but lost the guarantee of triviality. Since it seems like both (1) and (2) are really important, it seems like (3) is the only viable option for libc++, after all. This is also what other implementations are doing.
> 
> This patch moves libc++ from (1) to (3).
> 
> It also:
> 
> - Improves the test coverage for the various ways of initializing `std::array`
> - Adds tests for the triviality of `std::array`
> - Adds tests for the aggregate-ness of `std::array`

The requirement of uniqueness for return values is likely bogus, see [LWG2157](https://cplusplus.github.io/LWG/issue2157).

**Comment by frederick-vs-ja** [comment]
Perhaps we should ask @AlisdairM who made requirment of uniqueness between [N1479](https://wg21.link/n1479) and  [N1548](https://wg21.link/n1548) what are expected to be implemented.

**Comment by AlisdairM** [comment]
My original intent was to return a null pointer, but LWG expressed a desire that different empty array objects return different iterators, so you might catch the invalid case of comparing iterators from different containers.The idea at the time is that a viable implementation technique would be to return a reinterpret_cast of ‘this’; we did not intend to require a non-empty implementation just to implement the uniqueness property of iterators.AlisdairMSent from my iPhoneOn Dec 15, 2023, at 23:04, A. Jiang ***@***.***> wrote:﻿
Perhaps we should ask @AlisdairM who made requirment of uniqueness between N1479 and  N1548 what are expected to be implemented.

**Comment by ldionne** [comment]
Now, the issue with that is that using `reinterpret_cast` isn't `constexpr` friendly, which is why I switched it over to `nullptr`. To me, this sounds like LWG-issue material. I think the standard is impossible to implement correctly unless you're willing to give up the fact that `std::array<T, 0>` is a trivial type.

**Comment by mordante** [comment]
> Now, the issue with that is that using `reinterpret_cast` isn't `constexpr` friendly, which is why I switched it over to `nullptr`. To me, this sounds like LWG-issue material. I think the standard is impossible to implement correctly unless you're willing to give up the fact that `std::array<T, 0>` is a trivial type.

`std::bit_cast` is `constexpr` friendly. Its implementation uses `__builtin_bit_cast`. Could we use this builtin to fix this?

**Comment by ldionne** [comment]
`std::bit_cast` is not constexpr if either the `To` or the `From` type is a pointer.

**Comment by philnik777** [comment]
This is possible to implement with `_Tp __array[0]`, at least in ABIv2. I think it could also be implemented in the stable ABI.

**Comment by frederick-vs-ja** [comment]
> Now, the issue with that is that using `reinterpret_cast` isn't `constexpr` friendly, which is why I switched it over to `nullptr`.

It seems to me that we can make `array<T, 0>::iterator` distinct from other `array<T, N>::iterator`, and then `array<T, 0>::iterator::operator*` doesn't need to be `constexpr`-friendly because it's always UB to call such a `operator*`. We can even make `array<T, 0>::iterator::operator*` `constexpr` fucntions by some tricks avoiding IFNDR-ness (no longer needed since C++23 per [P2448R2](https://wg21.link/p2448r2)).

However, I believe it's better to render all of these tricks unnecessary by simply allowing `array<T, 0>::iterator::begin` to return a null value.

**Comment by AlisdairM** [comment]
As original author of the proposal I would be in favour of allowing (but not requiring?) a nullptr result.  I think the folks exploring the problems here are in a much better position to write up the LWG issue than I am though.Sent from my iPhoneOn Dec 18, 2023, at 21:28, A. Jiang ***@***.***> wrote:﻿

Now, the issue with that is that using reinterpret_cast isn't constexpr friendly, which is why I switched it over to nullptr.

It seems to me that we can make array<T, 0>::iterator distinct from other array<T, N>::iterator, and then array<T, 0>::iterator::operator* doesn't need to be constexpr-friendly because it's always UB to call such a operator*. We can even make array<T, 0>::iterator::operator* constexpr fucntions by some tricks avoiding IFNDR-ness (no longer needed since C++23 per P2448R2).
However, I believe it's better to render all of these tricks unnecessary by simply allowing array<T, 0>::iterator::begin to return a null value.
